astroid/interpreter/objectmodel.py

```diff
@@ -119,7 +119,35 @@ def lookup(self, name):
         raise AttributeInferenceError(target=self._instance, attribute=name)
 
 
-class ModuleModel(ObjectModel):
+class DunderCompletionMixin:
+    """Provide placeholders for further double-underscored methods for the
+    sake of completeness."""
+
+    @property
+    def attr___ne__(self):
+        return node_classes.Unknown()
+
+    attr___subclasshook__ = attr___ne__
```

Comment:
Ah, sorry, but I'm having some trouble following your comment. Can you explain what you mean? They all resolve to `Unknown` so I don't know that they do much besides prevent `no-member` errors and such.

Reply:
I'm not trying to be dense, I just think I'm still missing the point! By "come from" the mixin, you mean for someone reading astroid's source? And by coming from type or object in python, is that based on something other than the `__repr__`? I wouldn't put too much weight on the `__repr__`.

Reply:
I re-read https://docs.python.org/3/reference/datamodel.html to get to grips with it once again.

My thought process is that it's often better to try and replicate the data structure Python/CPython uses. Since most of these attributes are part of the description of base `objects` they should probably be defined on `ObjectModel`.

Put differently: `ObjectModel` should probably replicate `dir(object)` as all other models inherit from it and this would better represent how Python builds represents data models.

Reply:
Gotcha. So I just gave this a spin and got a raft of failures, see below for result.

Diff
```diff
diff --git a/ChangeLog b/ChangeLog
index 37b253d6..1c348ed1 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -6,8 +6,7 @@ What's New in astroid 2.12.0?
 =============================
 Release date: TBA
 
-* Created ``_DunderCompletionMixin`` in order to let ``ModuleModel`` share
-  the same dunder definitions that were already present on ``FunctionModel``.
+* Promoted most dunder definitions present on ``ModuleModel`` to ``ObjectModel``.
 
   Refs PyCQA/pylint#6094
 
diff --git a/astroid/interpreter/objectmodel.py b/astroid/interpreter/objectmodel.py
index 36717ae9..29a8d19c 100644
--- a/astroid/interpreter/objectmodel.py
+++ b/astroid/interpreter/objectmodel.py
@@ -118,15 +118,11 @@ class ObjectModel:
             return getattr(self, IMPL_PREFIX + name)
         raise AttributeInferenceError(target=self._instance, attribute=name)
 
-
-class _DunderCompletionMixin:
-    """Provide placeholders for further double-underscored methods for the
-    sake of completeness."""
-
     @property
     def attr___ne__(self):
         return node_classes.Unknown()
 
+    attr___doc__ = attr___ne__
     attr___subclasshook__ = attr___ne__
     attr___str__ = attr___ne__
     attr___sizeof__ = attr___ne__
@@ -135,6 +131,8 @@ class _DunderCompletionMixin:
     attr___reduce__ = attr___ne__
     attr___reduce_ex__ = attr___ne__
     attr___new__ = attr___ne__
+    attr___ge__ = attr___ne__
+    attr___le__ = attr___ne__
     attr___lt__ = attr___ne__
     attr___eq__ = attr___ne__
     attr___gt__ = attr___ne__
@@ -143,11 +141,12 @@ class _DunderCompletionMixin:
     attr___getattribute__ = attr___ne__
     attr___hash__ = attr___ne__
     attr___init__ = attr___ne__
+    attr___init_subclass__ = attr___ne__
     attr___dir__ = attr___ne__
     attr___class__ = attr___ne__
 
 
-class ModuleModel(ObjectModel, _DunderCompletionMixin):
+class ModuleModel(ObjectModel):
     def _builtins(self):
         builtins_ast_module = AstroidManager().builtins_module
         return builtins_ast_module.special_attributes.lookup("__dict__")
@@ -224,7 +223,7 @@ class ModuleModel(ObjectModel, _DunderCompletionMixin):
         return node_classes.Unknown()
 
 
-class FunctionModel(ObjectModel, _DunderCompletionMixin):
+class FunctionModel(ObjectModel):
     @property
     def attr___name__(self):
         return node_classes.Const(value=self._instance.name, parent=self._instance)
@@ -406,9 +405,9 @@ class FunctionModel(ObjectModel, _DunderCompletionMixin):
         return DescriptorBoundMethod(proxy=self._instance, bound=self._instance)
 
     # These are here just for completion.
-    attr___call__ = _DunderCompletionMixin.attr___ne__
-    attr___closure__ = _DunderCompletionMixin.attr___ne__
-    attr___code__ = _DunderCompletionMixin.attr___ne__
+    attr___call__ = ObjectModel.attr___ne__
+    attr___closure__ = ObjectModel.attr___ne__
+    attr___code__ = ObjectModel.attr___ne__
```

<details>
<summary> Result</summary>
=========================================================== FAILURES ===========================================================
_________________________________________ test_init_decorator_init_false[dataclasses] __________________________________________

module = 'dataclasses'

    @parametrize_module
    def test_init_decorator_init_false(module: str):
        """Test that no init is generated when init=False is passed to
        dataclass decorator.
        """
        node = astroid.extract_node(
            f"""
        from {module} import dataclass
        from typing import List
    
        @dataclass(init=False)
        class A:
            x: int
            y: str
            z: List[bool]
    
        A.__init__ #@
        """
        )
        init = next(node.infer())
>       assert init._proxied.parent.name == "object"
E       AssertionError: assert Uninferable == 'object'
E        +  where Uninferable = Uninferable.name
E        +    where Uninferable = Uninferable.parent
E        +      where Uninferable = Uninferable._proxied

tests/unittest_brain_dataclasses.py:536: AssertionError
_____________________________________ test_init_decorator_init_false[pydantic.dataclasses] _____________________________________

module = 'pydantic.dataclasses'

    @parametrize_module
    def test_init_decorator_init_false(module: str):
        """Test that no init is generated when init=False is passed to
        dataclass decorator.
        """
        node = astroid.extract_node(
            f"""
        from {module} import dataclass
        from typing import List
    
        @dataclass(init=False)
        class A:
            x: int
            y: str
            z: List[bool]
    
        A.__init__ #@
        """
        )
        init = next(node.infer())
>       assert init._proxied.parent.name == "object"
E       AssertionError: assert Uninferable == 'object'
E        +  where Uninferable = Uninferable.name
E        +    where Uninferable = Uninferable.parent
E        +      where Uninferable = Uninferable._proxied

tests/unittest_brain_dataclasses.py:536: AssertionError
____________________________________ test_init_decorator_init_false[marshmallow_dataclass] _____________________________________

module = 'marshmallow_dataclass'

    @parametrize_module
    def test_init_decorator_init_false(module: str):
        """Test that no init is generated when init=False is passed to
        dataclass decorator.
        """
        node = astroid.extract_node(
            f"""
        from {module} import dataclass
        from typing import List
    
        @dataclass(init=False)
        class A:
            x: int
            y: str
            z: List[bool]
    
        A.__init__ #@
        """
        )
        init = next(node.infer())
>       assert init._proxied.parent.name == "object"
E       AssertionError: assert Uninferable == 'object'
E        +  where Uninferable = Uninferable.name
E        +    where Uninferable = Uninferable.parent
E        +      where Uninferable = Uninferable._proxied

tests/unittest_brain_dataclasses.py:536: AssertionError
________________________________________________ test_invalid_init[dataclasses] ________________________________________________

module = 'dataclasses'

    @parametrize_module
    def test_invalid_init(module: str):
        """Test that astroid doesn't generate an initializer when attribute order is invalid."""
        node = astroid.extract_node(
            f"""
        from {module} import dataclass
    
        @dataclass
        class A:
            arg1: float = 0.0
            arg2: str
    
        A.__init__  #@
        """
        )
        init = next(node.infer())
>       assert init._proxied.parent.name == "object"
E       AssertionError: assert Uninferable == 'object'
E        +  where Uninferable = Uninferable.name
E        +    where Uninferable = Uninferable.parent
E        +      where Uninferable = Uninferable._proxied

tests/unittest_brain_dataclasses.py:652: AssertionError
___________________________________________ test_invalid_init[pydantic.dataclasses] ____________________________________________

module = 'pydantic.dataclasses'

    @parametrize_module
    def test_invalid_init(module: str):
        """Test that astroid doesn't generate an initializer when attribute order is invalid."""
        node = astroid.extract_node(
            f"""
        from {module} import dataclass
    
        @dataclass
        class A:
            arg1: float = 0.0
            arg2: str
    
        A.__init__  #@
        """
        )
        init = next(node.infer())
>       assert init._proxied.parent.name == "object"
E       AssertionError: assert Uninferable == 'object'
E        +  where Uninferable = Uninferable.name
E        +    where Uninferable = Uninferable.parent
E        +      where Uninferable = Uninferable._proxied

tests/unittest_brain_dataclasses.py:652: AssertionError
___________________________________________ test_invalid_init[marshmallow_dataclass] ___________________________________________

module = 'marshmallow_dataclass'

    @parametrize_module
    def test_invalid_init(module: str):
        """Test that astroid doesn't generate an initializer when attribute order is invalid."""
        node = astroid.extract_node(
            f"""
        from {module} import dataclass
    
        @dataclass
        class A:
            arg1: float = 0.0
            arg2: str
    
        A.__init__  #@
        """
        )
        init = next(node.infer())
>       assert init._proxied.parent.name == "object"
E       AssertionError: assert Uninferable == 'object'
E        +  where Uninferable = Uninferable.name
E        +    where Uninferable = Uninferable.parent
E        +      where Uninferable = Uninferable._proxied

tests/unittest_brain_dataclasses.py:652: AssertionError
_____________________________ NumpyBrainNdarrayTest.test_numpy_function_calls_inferred_as_ndarray ______________________________

self = <tests.unittest_brain_numpy_ndarray.NumpyBrainNdarrayTest testMethod=test_numpy_function_calls_inferred_as_ndarray>

    def test_numpy_function_calls_inferred_as_ndarray(self):
        """
        Test that some calls to numpy functions are inferred as numpy.ndarray
        """
        licit_array_types = ".ndarray"
        for func_ in self.ndarray_returning_ndarray_methods:
            with self.subTest(typ=func_):
                inferred_values = list(self._inferred_ndarray_method_call(func_))
                self.assertTrue(
                    len(inferred_values) == 1,
                    msg=f"Too much inferred value for {func_:s}",
                )
                self.assertTrue(
>                   inferred_values[-1].pytype() in licit_array_types,
                    msg=f"Illicit type for {func_:s} ({inferred_values[-1].pytype()})",
                )
E               TypeError: 'in <string>' requires string as left operand, not Uninferable

tests/unittest_brain_numpy_ndarray.py:142: TypeError
____________________________________________ InferenceTest.test__new__bound_methods ____________________________________________

self = <tests.unittest_inference.InferenceTest testMethod=test__new__bound_methods>

    def test__new__bound_methods(self) -> None:
        node = extract_node(
            """
        class cls(object): pass
        cls().__new__(cls) #@
        """
        )
        inferred = next(node.infer())
>       self.assertIsInstance(inferred, Instance)
E       AssertionError: Uninferable is not an instance of <class 'astroid.bases.Instance'>

tests/unittest_inference.py:1512: AssertionError
_____________________ FunctionModelTest.test_builtin_dunder_init_does_not_crash_when_accessing_annotations _____________________

self = <tests.unittest_object_model.FunctionModelTest testMethod=test_builtin_dunder_init_does_not_crash_when_accessing_annotations>

    def test_builtin_dunder_init_does_not_crash_when_accessing_annotations(
        self,
    ) -> None:
        ast_node = builder.extract_node(
            """
        class Class:
            @classmethod
            def class_method(cls):
                cls.__init__.__annotations__ #@
        """
        )
        inferred = next(ast_node.infer())
>       self.assertIsInstance(inferred, astroid.Dict)
E       AssertionError: Uninferable is not an instance of <class 'astroid.nodes.node_classes.Dict'>

tests/unittest_object_model.py:515: AssertionError
______________________________________ test_max_inferred_for_complicated_class_hierarchy _______________________________________

    def test_max_inferred_for_complicated_class_hierarchy() -> None:
        """Regression test for a crash reported in https://github.com/PyCQA/pylint/issues/5679.
    
        The class hierarchy of 'sqlalchemy' is so intricate that it becomes uninferable with
        the standard max_inferred of 100. We used to crash when this happened.
        """
        # Create module and get relevant nodes
        module = resources.build_file(
            str(resources.RESOURCE_PATH / "max_inferable_limit_for_classes" / "main.py")
        )
        init_attr_node = module.body[-1].body[0].body[0].value.func
        init_object_node = module.body[-1].mro()[-1]["__init__"]
        super_node = next(init_attr_node.expr.infer())
    
        # Arbitrarily limit the max number of infered nodes per context
        InferenceContext.max_inferred = -1
        context = InferenceContext()
    
        # Try to infer 'object.__init__' > because of limit is impossible
        for inferred in bases._infer_stmts([init_object_node], context, frame=super):
            assert inferred == Uninferable
    
        # Reset inference limit
        InferenceContext.max_inferred = 100
        # Check that we don't crash on a previously uninferable node
>       assert super_node.getattr("__init__", context=context)[0] == Uninferable
E       assert <Unknown.Unknown l.None at 0x1275fc5e0> == Uninferable

tests/unittest_regrtest.py:430: AssertionError
=================================================== short test summary info ====================================================
FAILED tests/unittest_brain_dataclasses.py::test_init_decorator_init_false[dataclasses] - AssertionError: assert Uninferable ...
FAILED tests/unittest_brain_dataclasses.py::test_init_decorator_init_false[pydantic.dataclasses] - AssertionError: assert Uni...
FAILED tests/unittest_brain_dataclasses.py::test_init_decorator_init_false[marshmallow_dataclass] - AssertionError: assert Un...
FAILED tests/unittest_brain_dataclasses.py::test_invalid_init[dataclasses] - AssertionError: assert Uninferable == 'object'
FAILED tests/unittest_brain_dataclasses.py::test_invalid_init[pydantic.dataclasses] - AssertionError: assert Uninferable == '...
FAILED tests/unittest_brain_dataclasses.py::test_invalid_init[marshmallow_dataclass] - AssertionError: assert Uninferable == ...
FAILED tests/unittest_brain_numpy_ndarray.py::NumpyBrainNdarrayTest::test_numpy_function_calls_inferred_as_ndarray - TypeErro...
FAILED tests/unittest_inference.py::InferenceTest::test__new__bound_methods - AssertionError: Uninferable is not an instance ...
FAILED tests/unittest_object_model.py::FunctionModelTest::test_builtin_dunder_init_does_not_crash_when_accessing_annotations
FAILED tests/unittest_regrtest.py::test_max_inferred_for_complicated_class_hierarchy - assert <Unknown.Unknown l.None at 0x12...
</details>

Reply:
```diff
diff --git a/astroid/bases.py b/astroid/bases.py
index 2fbb11d40..2eefae092 100644
--- a/astroid/bases.py
+++ b/astroid/bases.py
@@ -183,7 +183,9 @@ class BaseInstance(Proxy):
             values = self._proxied.instance_attr(name, context)
         except AttributeInferenceError as exc:
             if self.special_attributes and name in self.special_attributes:
-                return [self.special_attributes.lookup(name)]
+                values = [self.special_attributes.lookup(name)]
+                if not all(i.name == "Unknown" for i in values):
+                    return values
 
             if lookupclass:
                 # Class attributes not available through the instance
diff --git a/astroid/nodes/scoped_nodes/scoped_nodes.py b/astroid/nodes/scoped_nodes/scoped_nodes.py
index 2c11ac5b4..cff9d7b89 100644
--- a/astroid/nodes/scoped_nodes/scoped_nodes.py
+++ b/astroid/nodes/scoped_nodes/scoped_nodes.py
@@ -2552,6 +2552,12 @@ class ClassDef(mixins.FilterStmtsMixin, LocalsDictNodeNG, node_classes.Statement
             raise AttributeInferenceError(target=self, attribute=name, context=context)
 
         values = self.locals.get(name, [])
+
+        # don't modify the list in self.locals!
+        values = list(values)
+        for classnode in self.ancestors(recurs=True, context=context):
+            values += classnode.locals.get(name, [])
+
         if name in self.special_attributes and class_context and not values:
             result = [self.special_attributes.lookup(name)]
             if name == "__bases__":
@@ -2560,11 +2566,6 @@ class ClassDef(mixins.FilterStmtsMixin, LocalsDictNodeNG, node_classes.Statement
                 result += values
             return result
 
-        # don't modify the list in self.locals!
-        values = list(values)
-        for classnode in self.ancestors(recurs=True, context=context):
-            values += classnode.locals.get(name, [])
-
         if class_context:
             values += self._metaclass_lookup_attribute(name, context)
 
```

I only get one failed test with this diff. Based on the comments we might actually be okay with that one failing: we go from an `Uniferable` to a result, although I'd need some more investigation to see if this is indeed true.

The `diff` indicates that we kind of depend on the `ObjectModel` not being complete: we look up certain attributes on the model and if we can find them we return to our actual inference of `object`. For example: ``ObjectModel.__new__`` (or ``attr___new__``) doesn't exist so we infer ``__new__`` from the ``ClassDef`` of `object`. Perhaps by explicitly defining `attr___new` we don't even need this diff and can just rely on `ObjectModel.__new__` instead of the `ClassDef`?

Reply:
I follow this at a high level but not completely because I haven't stepped through it yet.

> Perhaps by explicitly defining attr___new we don't even need this diff and can just rely on ObjectModel.__new__ instead of the ClassDef?

I'm not sure exactly what this is--I think you're the better expert than I at this point. Would you like to take over this PR?

Reply:
I'll have a look at this somewhere this week as now I'm intrigued as to why this is failing.

I'll report back if I can't find the solution 😄